File: app/src/main/java/br/com/alura/leilao/ui/activity/LancesLeilaoActivity.java
package br.com.alura.leilao.ui.activity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import br.com.alura.leilao.R;
import br.com.alura.leilao.api.EnviadorDeLance;
import br.com.alura.leilao.api.retrofit.client.LeilaoWebClient;
import br.com.alura.leilao.database.dao.UsuarioDAO;
import br.com.alura.leilao.model.Lance;
import br.com.alura.leilao.model.Leilao;
import br.com.alura.leilao.ui.dialog.AvisoDialogManager;
import br.com.alura.leilao.ui.dialog.NovoLanceDialog;

import static br.com.alura.leilao.ui.activity.LeilaoConstantes.CHAVE_LEILAO;

public class LancesLeilaoActivity extends AppCompatActivity{

   private static final String TITULO_APPBAR = "Lances do leilão";
   private final LeilaoWebClient client = new LeilaoWebClient();
   private final AvisoDialogManager dialogManager = new AvisoDialogManager(this);
   private TextView maioresLances;
   private TextView menorLance;
   private TextView maiorLance;
   private TextView descricao;
   private UsuarioDAO dao;
   private Leilao leilaoRecebido;

   @Override
   protected void onCreate(Bundle savedInstanceState){
      super.onCreate(savedInstanceState);
      setContentView(R.layout.activity_lances_leilao);
      getSupportActionBar().setTitle(TITULO_APPBAR);
      carregaLeilaoRecebido();
   }

   private void carregaLeilaoRecebido(){
      Intent dadosRecebidos = getIntent();
      if(dadosRecebidos.hasExtra(CHAVE_LEILAO)){
         leilaoRecebido = (Leilao) dadosRecebidos.getSerializableExtra(CHAVE_LEILAO);
         inicializaAtributos();
         inicializaViews();
         apresentaDados();
         configuraFab();
      }
      else{
         finish();
      }
   }

   private void inicializaViews(){
      descricao = findViewById(R.id.lances_leilao_descricao);
      maiorLance = findViewById(R.id.lances_leilao_maior_lance);
      menorLance = findViewById(R.id.lances_leilao_menor_lance);
      maioresLances = findViewById(R.id.lances_leilao_maiores_lances);
   }

   private void inicializaAtributos(){
      dao = new UsuarioDAO(this);
   }

   private void configuraFab(){
      FloatingActionButton fabAdiciona = findViewById(R.id.lances_leilao_fab_adiciona);
      fabAdiciona.setOnClickListener(new View.OnClickListener(){
         @Override
         public void onClick(View view){
            mostraDialogNovoLance();
         }
      });
   }

   private void mostraDialogNovoLance(){
      NovoLanceDialog dialog = new NovoLanceDialog(
         this,
         new NovoLanceDialog.LanceCriadoListener(){
            @Override
            public void lanceCriado(Lance lance){
               enviaLance(lance);
            }
         },
         dao, dialogManager);
      dialog.mostra();
   }

   private void enviaLance(Lance lance){
      EnviadorDeLance enviador = new EnviadorDeLance(
         client,
         lanceProcessadoListener(),
         dialogManager);
      enviador.envia(leilaoRecebido, lance);
   }

   @NonNull
   private EnviadorDeLance.LanceProcessadoListener lanceProcessadoListener(){
      return new EnviadorDeLance.LanceProcessadoListener(){
         @Override
         public void processado(Leilao leilao){
            leilaoRecebido = leilao;
            apresentaDados();
         }
      };
   }

   private void apresentaDados(){
      adicionaDescricao(leilaoRecebido);
      adicionaMaiorLance(leilaoRecebido);
      adicionaMenorLance(leilaoRecebido);
      adicionaMaioresLances(leilaoRecebido);
   }

   private void adicionaMaioresLances(Leilao leilao){
      StringBuilder stringBuilder = new StringBuilder();
      if(leilao.getLances().get(0).getValor() != 0.0){
         for(Lance lance : leilao.getTresMaioresLances()){
            stringBuilder.append("- ");
            stringBuilder.append(lance.getValorFormatado());
            stringBuilder.append("\n");
         }
      }
      String maioresLancesEmTexto = stringBuilder.toString();
      maioresLances.setText(maioresLancesEmTexto);
   }

   private void adicionaMenorLance(Leilao leilao){
      menorLance.setText(leilao.getMenorLanceFormatado());
   }

   private void adicionaMaiorLance(Leilao leilao){
      maiorLance.setText(leilao.getMaiorLanceFormatado());
   }

   private void adicionaDescricao(Leilao leilao){
      descricao.setText(leilao.getDescricao());
   }
}

File: app/src/main/java/br/com/alura/leilao/ui/activity/LancesLeilaoActivity.kt
package br.com.alura.leilao.ui.activity

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import br.com.alura.leilao.api.EnviadorDeLance
import br.com.alura.leilao.api.retrofit.client.LeilaoWebClient
import br.com.alura.leilao.database.AppDatabase
import br.com.alura.leilao.databinding.ActivityLancesLeilaoBinding
import br.com.alura.leilao.model.Lance
import br.com.alura.leilao.model.Leilao
import br.com.alura.leilao.ui.dialog.AvisoDialogManager
import br.com.alura.leilao.ui.dialog.NovoLanceDialog

class LancesLeilaoActivity : AppCompatActivity() {

    private lateinit var binding: ActivityLancesLeilaoBinding
    private val client = LeilaoWebClient()
    private val dialogManager = AvisoDialogManager(this)
    private val dao by lazy {
        AppDatabase.getDatabase(this).usuarioDao()
    }
    private lateinit var leilaoRecebido: Leilao

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLancesLeilaoBinding.inflate(layoutInflater)
        setContentView(binding.root)
        supportActionBar?.title = "Lances do leilão"
        carregaLeilaoRecebido()
    }

    private fun carregaLeilaoRecebido() {
        val dadosRecebidos = intent
        if (dadosRecebidos.hasExtra(CHAVE_LEILAO)) {
            leilaoRecebido =
                dadosRecebidos.getSerializableExtra(CHAVE_LEILAO) as Leilao
            apresentaDados()
            configuraFab()
        } else {
            finish()
        }
    }

    private fun configuraFab() {
        binding.lancesLeilaoFabAdiciona.setOnClickListener {
            mostraDialogNovoLance()
        }
    }

    private fun mostraDialogNovoLance() {
        val dialog = NovoLanceDialog(
            this,
            object : NovoLanceDialog.LanceCriadoListener {
                override fun lanceCriado(lance: Lance) {
                    enviaLance(lance)
                }
            },
            dao, dialogManager
        )
        dialog.mostra()
    }

    private fun enviaLance(lance: Lance) {
        val enviador = EnviadorDeLance(
            client,
            lanceProcessadoListener(),
            dialogManager
        )
        enviador.envia(leilaoRecebido, lance)
    }

    private fun lanceProcessadoListener() = object : EnviadorDeLance.LanceProcessadoListener {
        override fun processado(leilao: Leilao) {
            leilaoRecebido = leilao
            apresentaDados()
        }
    }

    private fun apresentaDados() {
        adicionaDescricao(leilaoRecebido)
        adicionaMaiorLance(leilaoRecebido)
        adicionaMenorLance(leilaoRecebido)
        adicionaMaioresLances(leilaoRecebido)
    }

    private fun adicionaMaioresLances(leilao: Leilao) {
        val stringBuilder = StringBuilder()
        if (leilao.lances[0].valor != 0.0) {
            for (lance in leilao.getTresMaioresLances()) {
                stringBuilder.append("- ")
                stringBuilder.append(lance.valorFormatado)
                stringBuilder.append(" - (")
                stringBuilder.append(lance.usuario.id)
                stringBuilder.append(") ")
                stringBuilder.append(lance.usuario.nome)
                stringBuilder.append("\n")
            }
        }
        binding.lancesLeilaoMaioresLances.text = stringBuilder.toString()
    }

    private fun adicionaMenorLance(leilao: Leilao) {
        binding.lancesLeilaoMenorLance.text = leilao.getMenorLanceFormatado()
    }

    private fun adicionaMaiorLance(leilao: Leilao) {
        binding.lancesLeilaoMaiorLance.text = leilao.getMaiorLanceFormatado()
    }

    private fun adicionaDescricao(leilao: Leilao) {
        binding.lancesLeilaoDescricao.text = leilao.descricao
    }
}